src/common/transport/h5_transport.cpp

```diff
@@ -303,19 +303,20 @@ uint32_t H5Transport::send(const std::vector<uint8_t> &data) noexcept
 
         while (remainingRetransmissions--)
         {
-            logPacket(true, h5EncodedPacket);
-            const auto err_code = nextTransportLayer->send(lastPacket);
-
-            if (err_code != NRF_SUCCESS)
-                return err_code;
-
             uint8_t seqNumBefore;
 
             {
                 std::unique_lock<std::recursive_mutex> seqNumLck(seqNumMutex);
                 seqNumBefore = seqNum;
             }
 
+
+            logPacket(true, h5EncodedPacket);
+            const auto err_code = nextTransportLayer->send(lastPacket);
```

Comment:
On second thought, would it make sense to use std::atomic<uint8_t> for seq/ackNum instead? That way we can do away with locking altogether.

Again, this doesn't have to be done for this PR, but it's nice to have in the future.

Reply:
Thank you for your review @martingalvan-nordic ! I will add a comment to explain why `seqNumBefore` should be assigned before calling `send()`.

It has been a while since I last looked at this codebase but after skimming through the code I'm fairly sure the `seqNum` is manipulated only when the `ACK_PACKET` is being received.

You have good ideas for further improvements. I hope someone will catch and do the improvements in follow-up pr. Unfortunately I can't commit my time for further development. However, I'm happy to help testing at least if `pc-ble-driver-py` wheel packages are provided.

One thing to note about using `std::atomic<uint8_t>` and not locks at all is that in that case also the use of condition variable `ackReceived` needs to be revisited.

Reply:
Sure thing, there's no need to do this yourself :)

I'll take a look at your changes ASAP, thanks!